### PR TITLE
Fix STACObject inheritance

### PR DESCRIPTION
**Pull request by duckontheweb**
**Related Issue(s):** #

* #410 

**Description:**

Updates `from_dict` and `from_file` methods for `STACObject` and all sub-classes so that any classes inheriting from them will be returned by those methods. This makes it easier for downstream libraries like pystac-client to create custom sub-classes. See [these tests](https://github.com/duckontheweb/pystac/blob/6e0785b45960c178c7b73039220589ac6f4723ac/tests/test_catalog.py#L1112-L1133) for an example.

Also moves the logic from `serialization.stac_object_from_dict` into `stac_io` to simplify the chain of functions that is called. If it seems like downstream libraries are using this function directly we can go through a deprecation process, but it seemed unlikely. Once `STAC_IO` is removed in 1.0.0 this code will only be used in one place in the `StacIO` class, so it seemed to be simpler to just move it in there.

Supersedes #443 

**PR Checklist:**

- [x] Code is formatted (run `pre-commit run --all-files`)
- [x] Tests pass (run `scripts/test`)
- [ ] This PR maintains or improves overall codebase code coverage.
- [ ] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/develop/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

**Comment by duckontheweb** [comment]
@matthewhanson I believe this will close #410, but would like to get your confirmation that it covers the needs in that issue.

**Comment by lossyrob** [comment]
@duckontheweb  Want to clarify before review - can you talk about why STACObject.from_file uses logic that seems to be duplicated in StacIO.read_stac_object, when it was just a straight call before? It seems like this introduces that logic in a couple of different places (e.g. also `pystac.read_file`), where it could be consolidated into the StacIO method. I agree with the move out of `serialization` into StacIO, just not sure why the `identify`, `migrate` etc checks are duplicated in other places.

**Comment by duckontheweb** [comment]
> @duckontheweb Want to clarify before review - can you talk about why STACObject.from_file uses logic that seems to be duplicated in StacIO.read_stac_object, when it was just a straight call before? It seems like this introduces that logic in a couple of different places (e.g. also `pystac.read_file`), where it could be consolidated into the StacIO method. I agree with the move out of `serialization` into StacIO, just not sure why the `identify`, `migrate` etc checks are duplicated in other places.

Yeah, definitely a good question. `StacIO.read_stac_object` calls `StacIO.stac_object_from_dict`, which has hard-coded classes (`pystac.Catalog`, `pystac.Collection`, `pystac.Item`) that it calls `from_dict` methods on to get a `STACObject` sub-class instance. For custom subclasses (e.g. `class API(pystac.Catalog)`) this breaks the ability to get back the original calling class in `from_file`. 

I originally considered passing the `cls` argument from that method down through the chain, but it seemed like it was going to change the signatures of too many functions, and I didn't want to introduce that change. 

I agree the duplication of logic is not ideal, and makes maintenance trickier. I can take another look at this and see if there is a better solution that doesn't involve altering the `StacIO` method signatures and gives us some DRY-er code.

**Review by matthewhanson — APPROVED** [review]

**Comment by matthewhanson** [comment]
this looks good, I think the consolidating the duplicated functions addresses the concern @lossyrob had.

Will test this out with pystac-client in a few.

File: tests/test_writing.py [hunk 1]
```diff
@@ -72,7 +72,9 @@ def validate_catalog_link_type(
             href: str, link_type: str, should_include_self: bool
         ) -> None:
             cat_dict = pystac.StacIO.default().read_json(href)
-            cat = pystac.Catalog.from_file(href)
+            cat = pystac.read_file(href)
+            if not isinstance(cat, pystac.Catalog):
+                raise pystac.STACTypeError(f"File at {href} is not a Catalog.")
```
Comment (duckontheweb) [inline tests/test_writing.py:77]: @lossyrob @matthewhanson One consequence of adding the `STACObject.identify` methods is that `Catalog.from_file` will now fail on a Collection file, which is why these lines changed. Judging from discussion in radiantearth/stac-spec#1082 I think this is the right behavior, but let me know if you feel otherwise. If we feel it should succeed, I can update `Catalog.identify` accordingly.

Comment (matthewhanson) [reply tests/test_writing.py:77]: I think that's the right behavior. A Collection is not a Catalog, so should not be able to be opened up as such

**Review by lossyrob — APPROVED** [review]
Minor changes suggested, LGTM otherwise!

File: pystac/__init__.py [hunk 2]
```diff
@@ -94,7 +94,18 @@ def read_file(href: str) -> STACObject:
             a :class:`~pystac.STACObject` and must be read using
             :meth:`ItemCollection.from_file <pystac.ItemCollection.from_file>`
     """
-    return STACObject.from_file(href)
+    stac_io = StacIO.default()
```
[the remark this answers is not in the document]
Comment (duckontheweb) [reply pystac/__init__.py:98]: Yeah, I agree. It's probably good to allow users to pass in existing `StacIO` instances wherever possible.

File: pystac/catalog.py [hunk 3]
```diff
@@ -946,7 +953,16 @@ def full_copy(
...
+
         result = super().from_file(href, stac_io)
         if not isinstance(result, Catalog):
             raise pystac.STACTypeError(f"{result} is not a {Catalog}.")
+        result._stac_io = stac_io
+
         return result
+
+    @classmethod
+    def identify_dict(cls, d: Dict[str, Any]) -> bool:
```
[the remark this answers is not in the document]
Comment (duckontheweb) [reply pystac/catalog.py:967]: This is defined as an abstract method in `STACObject` so it would need to have a consistent naming convention across all sub-classes. I didn't really like `identify_dict` either, but I couldn't think of a better generic option. Maybe `validate_dict_type`?

[the remark this answers is not in the document]
Comment (duckontheweb) [reply pystac/catalog.py:967]: Although, now that I look at it, it doesn't need to be an abstract class because it's only used in specific sub-class method definitions. I'll update the name as you suggested.

[the remark this answers is not in the document]
Comment (duckontheweb) [reply pystac/catalog.py:967]: Ah, now I remember why I did it this way. I wanted to standardize the name so that a sub-class like `pystac_client.API` could overwrite that method and have the name still make sense.

Something like:

```python
class API(pystac.Catalog):
    def identify_dict(cls, d):
        return super().identify_dict() and "conformance" in d
```

maybe `dict_matches_object_type` would be better.

[the remark this answers is not in the document]
Comment (lossyrob) [reply pystac/catalog.py:967]: Makes sense - perhaps `matches_object_type` or `is_object_type`, as the `dict` is in the method signature?

File: pystac/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (duckontheweb) [reply pystac/__init__.py]: Fixed in 4ebc6609902a0d6b64bc19a59a530376b0157a1e

File: pystac/catalog.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (duckontheweb) [reply pystac/catalog.py]: Renamed in e15c7c04d5904f144050285cf8d453fe8ea701af

**Comment by lossyrob** [comment]
@duckontheweb mind if I merge? Some changes I'm working on now for #453  should be based on these changes.

**Comment by duckontheweb** [comment]
No, go for it!

File: pystac/stac_object.py [hunk 4]
```diff
@@ -463,13 +463,17 @@ def from_file(
...
+
         if stac_io is None:
             stac_io = pystac.StacIO.default()
 
         if not is_absolute_href(href):
             href = make_absolute_href(href)
 
-        o = stac_io.read_stac_object(href)
+        d = stac_io.read_json(href)
+        o = cls.from_dict(d, href=href, migrate=True)
```
Comment (lossyrob) [inline pystac/stac_object.py:476]: Why this change? Couldn't this remain `stac_io.read_stac_object` for one codepath to handle href -> STACObject logic?

The type will be `STACObject` as far as type hinting goes, but the underlying class will be either a Catalog, Collection or Item. The casting done in the implementations should work out fine.

**Review by lossyrob — CHANGES_REQUESTED** [review]
I actually just merged in the changes from this branch, and had a question about STACObject.from_file, comment inline

File: pystac/stac_object.py (hunk 4 above)
Comment (lossyrob) [reply pystac/stac_object.py:476]: Just made this change and saw that some tests break with custom STACObjects. disregard.

**Review by lossyrob — APPROVED** [review]
